src/fastalloc/lru.rs

```diff
@@ -0,0 +1,331 @@
+use crate::{PReg, PRegSet, RegClass};
+use alloc::vec;
+use alloc::vec::Vec;
+use core::{
+    fmt,
+    ops::{Index, IndexMut},
+};
+use hashbrown::HashSet;
+
+const DUMMY_NODE_INDEX: usize = PReg::MAX + 1;
+
+/// A least-recently-used cache organized as a linked list based on a vector.
+pub struct Lru {
+    /// The list of node information.
+    ///
+    /// Each node corresponds to a physical register.
+    /// The index of a node is the `address` from the perspective of the linked list.
+    pub data: Vec<LruNode>,
+    /// Index of the most recently used register.
+    pub head: u8,
+    /// Class of registers in the cache.
+    pub regclass: RegClass,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct LruNode {
+    /// The previous physical register in the list.
+    pub prev: u8,
+    /// The next physical register in the list.
+    pub next: u8,
+}
+
+impl Lru {
+    pub fn new(regclass: RegClass, regs: &[PReg]) -> Self {
+        let mut data = vec![
+            LruNode {
+                prev: u8::MAX,
+                next: u8::MAX
+            };
+            PReg::MAX + 1
+        ];
+        let no_of_regs = regs.len();
+        for i in 0..no_of_regs {
+            let (reg, prev_reg, next_reg) = (
+                regs[i],
+                regs[i.checked_sub(1).unwrap_or(no_of_regs - 1)],
+                regs[if i >= no_of_regs - 1 { 0 } else { i + 1 }],
+            );
+            data[reg.hw_enc()].prev = prev_reg.hw_enc() as u8;
```

Comment:
`hw_enc` is used because the LRUs are parted by regclass. A single LRU is for a single class.

Reply:
Sure, but the index still fits in a `u8` and using the index avoids conversions between hw_enc and PReg (which just holds the index internally).

Reply:
The index also has the regclass in its lower bits which changes the value of the `hw_enc`

Reply:
Indeed, the issue is that this will result in sparse usage of `data` if we use the bitpacked index -- not the worst problem but something it'd be reasonable to avoid if we can. I'm personally fine with `hw_enc` here.